### DISCUSSION TITLE
Crash on startup since 0.24.0

### DISCUSSION THREAD

**Discussion by mackoj** [pointfreeco/swift-composable-architecture#752] [General]
Hi,

**Describe the bug**

Since [this modification](https://github.com/pointfreeco/swift-composable-architecture/commit/a38e04da7d665ed27ef0dbcfd3b8096a374c3fdb#diff-52e8a72d908dc1b0d23451f4b256d0244f27c6361facaa5cf73a026f315c25df) my app crash on device all the time on startup [here](https://github.com/pointfreeco/swift-composable-architecture/blob/682e2902c276afc200f7109fc09c3c19841b70b3/Sources/ComposableArchitecture/ViewStore.swift#L82).

I had to revert to an old version of TCA(0.23.0) to prevent the bug.

Thread 1: EXC_BAD_ACCESS (code=2, address=0x16f8a7ff8)

```
Thread 1 Queue : com.apple.main-thread (serial)
#0	0x00000001f010179c in ___chkstk_darwin ()
#1	0x000000010152fce0 in assignWithCopy for AppState ()
#2	0x00000001b84baed4 in specialized CurrentValueSubject.value.setter ()
#3	0x00000001b84b7fe4 in CurrentValueSubject.value.setter ()
#4	0x00000001033725e4 in closure #1 in ViewStore.init(_:removeDuplicates:) at /Users/jeffreymacko/Library/Developer/Xcode/DerivedData/sambot-fwigkzjcripmhnesyjiuzmjrzoyz/SourcePackages/checkouts/swift-composable-architecture/Sources/ComposableArchitecture/ViewStore.swift:82
#5	0x00000001b8424620 in Subscribers.Sink.receive(_:) ()
#6	0x00000001b8424da4 in protocol witness for Subscriber.receive(_:) in conformance Subscribers.Sink<τ_0_0, τ_0_1> ()
#7	0x00000001b84118f0 in FilterProducer.receive(_:) ()
#8	0x00000001b8412550 in protocol witness for Subscriber.receive(_:) in conformance FilterProducer<τ_0_0, τ_0_1, τ_0_2, τ_0_3, τ_0_4> ()
#9	0x00000001b84ba43c in CurrentValueSubject.Conduit.request(_:) ()
#10	0x00000001b84811b8 in protocol witness for Subscription.request(_:) in conformance ConduitBase<τ_0_0, τ_0_1> ()
#11	0x00000001b8410d64 in FilterProducer.request(_:) ()
#12	0x00000001b8412598 in protocol witness for Subscription.request(_:) in conformance FilterProducer<τ_0_0, τ_0_1, τ_0_2, τ_0_3, τ_0_4> ()
#13	0x00000001b84245c4 in Subscribers.Sink.receive(subscription:) ()
#14	0x00000001b8424d80 in protocol witness for Subscriber.receive(subscription:) in conformance Subscribers.Sink<τ_0_0, τ_0_1> ()
#15	0x00000001b841100c in FilterProducer.receive(subscription:) ()
#16	0x00000001b841252c in protocol witness for Subscriber.receive(subscription:) in conformance FilterProducer<τ_0_0, τ_0_1, τ_0_2, τ_0_3, τ_0_4> ()
#17	0x00000001b84b8e94 in CurrentValueSubject.receive<τ_0_0>(subscriber:) ()
#18	0x00000001b84ba784 in protocol witness for Publisher.receive<τ_0_0>(subscriber:) in conformance CurrentValueSubject<τ_0_0, τ_0_1> ()
#19	0x00000001b845d248 in Publisher.subscribe<τ_0_0>(_:) ()
#20	0x00000001b8428fd8 in Publishers.RemoveDuplicates.receive<τ_0_0>(subscriber:) ()
#21	0x00000001b845d248 in Publisher.subscribe<τ_0_0>(_:) ()
#22	0x00000001b8424064 in Publisher<>.sink(receiveValue:) ()
#23	0x0000000103372314 in ViewStore.init(_:removeDuplicates:) at /Users/jeffreymacko/Library/Developer/Xcode/DerivedData/sambot-fwigkzjcripmhnesyjiuzmjrzoyz/SourcePackages/checkouts/swift-composable-architecture/Sources/ComposableArchitecture/ViewStore.swift:79
#24	0x0000000103371ef0 in ViewStore.__allocating_init(_:removeDuplicates:) ()
#25	0x0000000103373ad4 in ViewStore<>.init(_:) at /Users/jeffreymacko/Library/Developer/Xcode/DerivedData/sambot-fwigkzjcripmhnesyjiuzmjrzoyz/SourcePackages/checkouts/swift-composable-architecture/Sources/ComposableArchitecture/ViewStore.swift:277
#26	0x0000000100465504 in SambotAppDelegate.application(_:willFinishLaunchingWithOptions:) at /Users/jeffreymacko/code/jayandjay/sambot/App/iOS/AppDelegate.swift:27
#27	0x0000000100466074 in @objc SambotAppDelegate.application(_:willFinishLaunchingWithOptions:) ()
#28	0x00000001a6d5b8f0 in -[UIApplication _handleDelegateCallbacksWithOptions:isSuspended:restoreState:] ()
#29	0x00000001a6d5d414 in -[UIApplication _callInitializationDelegatesWithActions:forCanvas:payload:fromOriginatingProcess:] ()
#30	0x00000001a6d63560 in -[UIApplication _runWithMainScene:transitionContext:completion:] ()
#31	0x00000001a63b71e4 in -[_UISceneLifecycleMultiplexer completeApplicationLaunchWithFBSScene:transitionContext:] ()
#32	0x00000001a6d5f30c in -[UIApplication _compellApplicationLaunchToCompleteUnconditionally] ()
#33	0x00000001a6d5f738 in -[UIApplication _run] ()
#34	0x00000001a6d64fcc in UIApplicationMain ()
#35	0x00000001ab24c380 in closure #1 in KitRendererCommon(_:) ()
#36	0x00000001ab24c30c in runApp<τ_0_0>(_:) ()
#37	0x00000001aad44b38 in static App.main() ()
#38	0x00000001004725e4 in static SambotApp.$main() at /Users/jeffreymacko/code/jayandjay/sambot/App/iOS/App.swift:8
#39	0x0000000100473004 in main ()
#40	0x00000001a3f9dcf8 in start ()
```

**Environment**

  - swift-composable-architecture version 0.25.1
  - Xcode 12.5.1 (12E507)
  - Swift 5.4
  - OS: iOS 14.7.1

**Additional context**

I did revert https://github.com/pointfreeco/swift-composable-architecture/commit/a38e04da7d665ed27ef0dbcfd3b8096a374c3fdb#diff-52e8a72d908dc1b0d23451f4b256d0244f27c6361facaa5cf73a026f315c25df to `.send($0)` in the 0.25.1 and it's not working 🤣.

**Comment by mbrandonw** [comment]
Sorry about that! 

> I did revert [a38e04d#diff-52e8a72d908dc1b0d23451f4b256d0244f27c6361facaa5cf73a026f315c25df](https://github.com/pointfreeco/swift-composable-architecture/commit/a38e04da7d665ed27ef0dbcfd3b8096a374c3fdb#diff-52e8a72d908dc1b0d23451f4b256d0244f27c6361facaa5cf73a026f315c25df) to `.send($0)` in the 0.25.1 and it's not working 🤣.

Can you explain this a bit more? Changing back to `.send` does not fix your issue? Are you sure that PR is what introduced the problem?

**Comment by mackoj** [comment]
I'm not 100% sure that this code https://github.com/pointfreeco/swift-composable-architecture/commit/a38e04da7d665ed27ef0dbcfd3b8096a374c3fdb#diff-52e8a72d908dc1b0d23451f4b256d0244f27c6361facaa5cf73a026f315c25df is the issue but when the app crash it stop there and there is modification here between the 0.23.0 and the 0.24.0.

When I do compare both https://github.com/pointfreeco/swift-composable-architecture/compare/0.23.0...0.24.0 I don't see what else could cause that crash.

What can I do to help?

This is just before the crash

<img width="1125" alt="Capture d’écran 2021-08-25 à 17 05 54" src="https://user-images.githubusercontent.com/661647/130816036-46c31343-5a6f-4745-9c30-0d12cb4a35ff.png">

**Comment by mackoj** [comment]
> Sorry about that!
> 
> > I did revert [a38e04d#diff-52e8a72d908dc1b0d23451f4b256d0244f27c6361facaa5cf73a026f315c25df](https://github.com/pointfreeco/swift-composable-architecture/commit/a38e04da7d665ed27ef0dbcfd3b8096a374c3fdb#diff-52e8a72d908dc1b0d23451f4b256d0244f27c6361facaa5cf73a026f315c25df) to `.send($0)` in the 0.25.1 and it's not working 🤣.
> 
> Can you explain this a bit more? Changing back to `.send` does not fix your issue? Are you sure that PR is what introduced the problem?

After bisecting for some time, I found out that my app was crash on every version for the last two months. But the archived version on TestFlight where working, the difference was the dependencies then. So I did retry to bisect and change TCA to a fixed version(0.23.0) and the app stopped crashing.

**Comment by andreyz** [comment]
Looks like the case of running out of stack space, judging by ``` #0	0x00000001f010179c in ___chkstk_darwin () ```
Two hunches: your AppState is too large or somehow you're stuck in a recursive update.

**Comment by mackoj** [comment]
Export of the thread and the variable with TCA 0.23.0

```
Thread 1 Queue : com.apple.main-thread (serial)
#0	0x00000001063a86b0 in closure #1 in ViewStore.init(_:removeDuplicates:) at /Users/jeffreymacko/Library/Developer/Xcode/DerivedData/sambot-fwigkzjcripmhnesyjiuzmjrzoyz/SourcePackages/checkouts/swift-composable-architecture/Sources/ComposableArchitecture/ViewStore.swift:82
#1	0x00000001b8424620 in Subscribers.Sink.receive(_:) ()
#2	0x00000001b8424da4 in protocol witness for Subscriber.receive(_:) in conformance Subscribers.Sink<τ_0_0, τ_0_1> ()
#3	0x00000001b84118f0 in FilterProducer.receive(_:) ()
#4	0x00000001b8412550 in protocol witness for Subscriber.receive(_:) in conformance FilterProducer<τ_0_0, τ_0_1, τ_0_2, τ_0_3, τ_0_4> ()
#5	0x00000001b84ba43c in CurrentValueSubject.Conduit.request(_:) ()
#6	0x00000001b84811b8 in protocol witness for Subscription.request(_:) in conformance ConduitBase<τ_0_0, τ_0_1> ()
#7	0x00000001b8410d64 in FilterProducer.request(_:) ()
#8	0x00000001b8412598 in protocol witness for Subscription.request(_:) in conformance FilterProducer<τ_0_0, τ_0_1, τ_0_2, τ_0_3, τ_0_4> ()
#9	0x00000001b84245c4 in Subscribers.Sink.receive(subscription:) ()
#10	0x00000001b8424d80 in protocol witness for Subscriber.receive(subscription:) in conformance Subscribers.Sink<τ_0_0, τ_0_1> ()
#11	0x00000001b841100c in FilterProducer.receive(subscription:) ()
#12	0x00000001b841252c in protocol witness for Subscriber.receive(subscription:) in conformance FilterProducer<τ_0_0, τ_0_1, τ_0_2, τ_0_3, τ_0_4> ()
#13	0x00000001b84b8e94 in CurrentValueSubject.receive<τ_0_0>(subscriber:) ()
#14	0x00000001b84ba784 in protocol witness for Publisher.receive<τ_0_0>(subscriber:) in conformance CurrentValueSubject<τ_0_0, τ_0_1> ()
#15	0x00000001b845d248 in Publisher.subscribe<τ_0_0>(_:) ()
#16	0x00000001b8428fd8 in Publishers.RemoveDuplicates.receive<τ_0_0>(subscriber:) ()
#17	0x00000001b845d248 in Publisher.subscribe<τ_0_0>(_:) ()
#18	0x00000001b8424064 in Publisher<>.sink(receiveValue:) ()
#19	0x00000001063a8440 in ViewStore.init(_:removeDuplicates:) at /Users/jeffreymacko/Library/Developer/Xcode/DerivedData/sambot-fwigkzjcripmhnesyjiuzmjrzoyz/SourcePackages/checkouts/swift-composable-architecture/Sources/ComposableArchitecture/ViewStore.swift:78
#20	0x00000001063a801c in ViewStore.__allocating_init(_:removeDuplicates:) ()
#21	0x00000001063a9b9c in ViewStore<>.init(_:) at /Users/jeffreymacko/Library/Developer/Xcode/DerivedData/sambot-fwigkzjcripmhnesyjiuzmjrzoyz/SourcePackages/checkouts/swift-composable-architecture/Sources/ComposableArchitecture/ViewStore.swift:252
#22	0x0000000104a02a04 in SambotAppDelegate.application(_:willFinishLaunchingWithOptions:) at /Users/jeffreymacko/code/jayandjay/sambot/App/iOS/AppDelegate.swift:27
#23	0x0000000104a03574 in @objc SambotAppDelegate.application(_:willFinishLaunchingWithOptions:) ()
#24	0x00000001a6d5b8f0 in -[UIApplication _handleDelegateCallbacksWithOptions:isSuspended:restoreState:] ()
#25	0x00000001a6d5d414 in -[UIApplication _callInitializationDelegatesWithActions:forCanvas:payload:fromOriginatingProcess:] ()
#26	0x00000001a6d63560 in -[UIApplication _runWithMainScene:transitionContext:completion:] ()
#27	0x00000001a63b71e4 in -[_UISceneLifecycleMultiplexer completeApplicationLaunchWithFBSScene:transitionContext:] ()
#28	0x00000001a6d5f30c in -[UIApplication _compellApplicationLaunchToCompleteUnconditionally] ()
#29	0x00000001a6d5f738 in -[UIApplication _run] ()
#30	0x00000001a6d64fcc in UIApplicationMain ()
#31	0x00000001ab24c380 in closure #1 in KitRendererCommon(_:) ()
#32	0x00000001ab24c30c in runApp<τ_0_0>(_:) ()
#33	0x00000001aad44b38 in static App.main() ()
#34	0x0000000104a0db98 in static SambotApp.$main() at /Users/jeffreymacko/code/jayandjay/sambot/App/iOS/App.swift:8
#35	0x0000000104a0e5b8 in main ()
#36	0x00000001a3f9dcf8 in start ()
```

<img width="1125" alt="Capture d’écran 2021-08-25 à 17 20 43" src="https://user-images.githubusercontent.com/661647/130818610-b38266b4-4e97-4939-ac61-c1cdebb034b8.png">

**Comment by andreyz** [comment]
Call me crazy, maybe the change you're [referring to](https://github.com/pointfreeco/swift-composable-architecture/commit/a38e04da7d665ed27ef0dbcfd3b8096a374c3fdb#diff-52e8a72d908dc1b0d23451f4b256d0244f27c6361facaa5cf73a026f315c25df) @mackoj is adding an additional copy (`assignWithCopy`) of your AppState, which causes running out of stack space...
```
#0	0x00000001f010179c in ___chkstk_darwin ()
#1	0x000000010152fce0 in assignWithCopy for AppState ()
```

**Comment by mackoj** [comment]
> Looks like the case of running out of stack space, judging by `#0 0x00000001f010179c in ___chkstk_darwin ()`
> Two hunches: your AppState is too large or somehow you're stuck in a recursive update.

Thanks, for responding.
I don't think it's a recursive state because the same code works fine with TCA 0.23.0.
It might be a stack space issue which would mean that TCA 0.24.0 takes more stack space than TCA 0.23.0. I will try this by reducing the size of the AppState and see if something changes.

**Comment by mackoj** [comment]
> Call me crazy, maybe the change you're referring to @mackoj is adding an additional copy (`assignWithCopy`) of your AppState, which causes running out of stack space...
> 
> ```
> #0	0x00000001f010179c in ___chkstk_darwin ()
> #1	0x000000010152fce0 in assignWithCopy for AppState ()
> ```

Great catch @andreyz for sure this https://github.com/pointfreeco/swift-composable-architecture/commit/a38e04da7d665ed27ef0dbcfd3b8096a374c3fdb#diff-52e8a72d908dc1b0d23451f4b256d0244f27c6361facaa5cf73a026f315c25df (diff of **ViewStore** between 0.23 and 0.24) show the copy of the app state.

And create the crash 😆

**Reply by mbrandonw** [reply]
I'm not sure I see why doing `.value = ...` would cause a copy whereas doing `.send(...)` does not. Are you certain this is the culprit? I thought you said you had tried to revert that one line and it didn't fix your crash?

It may be a pain to do but if you want to find the exact TCA commit that introduced this problem for you, you may need to do a manual "bisect" where you just pin to lots of TCA commits until you find the problematic one.

FWIW the only recent change I can think of that may have introduced another copy is when we coalesced multiple state updates into a single one:

https://github.com/pointfreeco/swift-composable-architecture/commit/3ca8249bd3262e9837c6cb4d352cc58e370e33fc#diff-50345ad35b59a32af74fda59695df9c6b1826302d5b41e5d8fbe47477917cc5dR373

However, that change has been in TCA since 0.20.0 :/

**Reply by mackoj** [reply]
> Are you certain this is the culprit?

I think so.

> I thought you said you had tried to revert that one line and it didn't fix your crash?

Yes I did say that but I think in the meantime maybe TCA stack utilisation might have grown up a little and since I am already at 20 KB of stack utilisation with a little more utilisation I might been over it.

> It may be a pain to do but if you want to find the exact TCA commit that introduced this problem for you, you may need to do a manual "bisect" where you just pin to lots of TCA commits until you find the problematic one.

I can will do it just to be sure.

**Reply by mbrandonw** [reply]
Sounds good, very interested to see what you find! If it really is that one line change then we'd be happy to revert.

**Comment by andreyz** [comment]
@mackoj I was a [little suspicious](https://github.com/pointfreeco/swift-composable-architecture/pull/713#discussion_r688131286) of the above change from `send` to assignment from the start. Now when I saw your stack trace, it suddenly made sense 💡

**Comment by mackoj** [comment]
I do think that revert the `send` is a first step to reduce the probability of this issue to happen to other but in my case I might need to reduce the size of my AppState anyway.

**Comment by tgrapperon** [comment]
StackOverflow crashes are the worst to spot. Usually EXC_BAD_ACCESS, code=2 with non-nil addresses at random places where it can't crash.
@mackoj You can box parts of your state in reference-type wrappers to put them on the heap. Copy-on-write property wrappers are often doing this for free as a side effect.

**Comment by mackoj** [comment]
> StackOverflow crashes are the worst to spot. Usually EXC_BAD_ACCESS, code=2 with non-nil addresses at random places where it can't crash.
> @mackoj You can box parts of your structs in reference-type wrappers to put them on the heap. Copy-on-write property wrapper are often doing this for free as a side effect.

I can solve it this way but I do think that AppState should be on the heap([stephencelis was the first to talk about this](https://github.com/pointfreeco/swift-composable-architecture/discussions/488)).

It did happen to me on various occasions and be face with this issue and each time it's pretty time-consuming. Find the root cause and the conclusion has often been removed feature or add `@Cow` until it works.

We cannot rely on the hard limit imposed by the OS since our AppState could go over them so easily.

What do you think? @tgrapperon @andreyz

FWIW I have a hard time foresee what issue putting AppState on the heap could bring.

**Comment by tgrapperon** [comment]
It's an ongoing issue with Swift that switches paradigms. The runtime hasn't caught up yet and 512kB is often too small, I agree. I think it should be around 5-10MB, like it is on macOS, especially with the new concurrency model that will limit the number of threads.
Fortunately, `Array`, `Dictionaries` and most of `String` are already on the heap, so it mitigates the vast majority of outgrowing states.

However, if you're using a `@CoW` property wrapper at the root of the state in the store, the state is still a value type itself, and it may overflow if you extract-it from the wrapper to pass-it to a function for example.
I guess in this case, it would be some kind of premature optimization that will only make the app crash somewhere else.
In other words, the only solution seems to have a state that is already manageable by itself, regardless of the context it is used/stored. It means putting branches on the heap manually, if needed.

Anecdotally, this is not proper to TCA. I had this kind of issue in my vanilla UIKit app where I have a value with many nested value fields. I had to turn the whole thing into a class because I still had the issue when extracting parts to pass to functions. Now that we have property-wrappers in function's arguments, I would try to push the `@CoW` experiment farther.

**Reply by mackoj** [reply]
Thanks for the explainer.

Since we cannot assume that the is limit will change. Do you think there is a reasonable path to solve this issue ?

**Reply by tgrapperon** [reply]
To my knowledge, there is no silver bullet.

If you reverse the issue, the state grows on the stack when you add a new value field. If this field's size on the stack is small (a number, a boolean, some container that stores on the heap like arrays, dictionaries or strings, a CGPoint, etc.), this is not problematic.
If the field is a container itself, like the state of a feature in TCA, I may consider wrapping it at some point, especially near the root state.

In the past, I've experimented with StateBindings (#451). Maybe it can be worthwhile to revisit this and automatically put the substate on the heap as we derive it. I have a private implementation using protocols that is faster than the purely functional approach presented in the PR. If there is traction, I may consider open-sourcing it.

**Comment by mbrandonw** [comment]
I'm going to convert this to a discussion so that it's easier for us to thread messages because there's a lot to discuss here.

**Comment by mackoj** [comment]
A bit more context informations the **AppState** is 20 KB
 
<details>

```swift
func bytesFormater(bytes: Int) -> String {
  let bf = ByteCountFormatter()
  bf.allowedUnits = .useAll
  bf.countStyle = .file
  bf.includesUnit = true
  bf.isAdaptive = true
  return bf.string(fromByteCount: Int64(bytes))
}

print("💾 AppState size: \(bytesFormater(bytes: MemoryLayout<AppState>.size))")
```
</details>

**Reply by stephencelis** [reply]
@mackoj Have you seen this discussion? https://github.com/pointfreeco/swift-composable-architecture/discussions/488

Has some fixes using a boxed type.

Edit: ah you linked to it in another post :)

**Reply by mackoj** [reply]
Using `@Cow` from [here](https://github.com/pointfreeco/swift-composable-architecture/discussions/488#discussioncomment-591715) and it reduce the AppState down to:
- 12 KB with TCA 0.23.0 and the app seems to work 🎉
- 13 KB with TCA 0.25.1 and the app seems to work 🎉

**Reply by mackoj** [reply]
I did try to reproduce the bug in a sample project without success. I will next try bisect on each commit to confirm the source of the issue.

[TestCurrentValueSubject.zip](https://github.com/pointfreeco/swift-composable-architecture/files/7050259/TestCurrentValueSubject.zip)

But from what I understand from the sample projet it might be something else at play here 🤷‍♂️

**Reply by iampatbrown** [reply]
While benchmarking I noticed `.send(...)` and `.value = ...` compiled slightly differently, however, I'm unsure how much this is impacting your particular situation. If the difference between the two is causing the app to crash, `AppState` is probably already too big. 

If you can get a failing example that would be awesome. I'd be keen to check it out.

**Comment by lukeredpath** [comment]
I’ve been seeing this crash too trying to upgrade our app from 0.23 to 0.25. It instantly started crashing on launch. 

However, we have some threading issues in our app - a legacy queue based effects system was accessing our TCA store concurrently from background threads and of course TCA isn’t thread safe. Fortunately we rarely saw memory related crashes until I attempted this upgrade. 

I’ve gone down the route of eliminating all cross thread store access so we only access the store from the main queue and the problem seems to have gone away. I don’t know why this change seemed to make it more prone to crashing when accessing across threads though. 

Are you sure you are only accessing the TCA store from a single thread?

**Reply by mackoj** [reply]
I will look into this too thanks 🙏

**Reply by mackoj** [reply]
You were right the source of the bug ~is~ was a threading issue.

Thanks,

**Comment by DaanyalMeyer** [comment]
We just started testing our app in iOS 16 and these stack overflow seem to happen all the time.

**Reply by mackoj** [reply]
This is a bad news I will try my app on iOS 16.

**Comment by tgrapperon** [comment]
As it seems to be an ongoing issue, I've drafted a few diagnostic tools this morning. Disclaimer: I'm not a specialist with low-level things, so please consider them with a pinch of salt.

### `StackState.print()` and `@PrintStackState`:
This prints the current thread's stack state: how much it is filled, and how many bytes are available.
Exact values are not important, but if it shows that the stack was full at 99% when it crashes, it is probably a stack overflow.
You can call `StackState.print()` anywhere, and it will display information like:
```swift
Stack state: using 29872 bytes out of 8372224, 0.36% full.
```
Alternatively, you can use the `@PrintStackState` property wrapper to make it print the same information each time you create, access or set the property.

### `@PrintSize`:
This is another property wrapper that you can put in front of a property to print the size of the wrapped value. You can also compose them: `@State @PrintSize @PrintStackState var value = 0`.

### `.debugStack()`
This higher order reducer prints the state's size once, and then the stack state at each action received.

This can most likely be improved, but feel free to tell me what you think!

---
```swift
@propertyWrapper
public struct PrintSize<Value> {
  public init(wrappedValue: Value, label: String? = nil) {
    self.wrappedValue = wrappedValue
    print("""
    Size of \(Value.self)\(label.map {"(\($0))"} ?? ""): \
    \(MemoryLayout.size(ofValue: wrappedValue)) bytes.
    """)
  }
  public var wrappedValue: Value
}

@propertyWrapper
public struct PrintStackState<Value> {
  public init(wrappedValue: Value, label: String? = nil) {
    self._wrappedValue = wrappedValue
    self.label = label
    print(formatOuput(prefix: "Init "))
  }
  private var _wrappedValue: Value
  var label: String?
  public var wrappedValue: Value {
    get {
      print(formatOuput(prefix: "Get "))
      return _wrappedValue
    }
    set {
      print(formatOuput(prefix: "Set "))
      _wrappedValue = newValue
    }
  }
  func formatOuput(prefix: String) -> String {
    "\(prefix)\(Value.self)\(label.map {"(\($0)):"} ?? "")\(StackState.current)"
  }
}

public struct StackState: CustomDebugStringConvertible {
  public static var current: Self { .init() }
  public static func print() { Swift.print(current) }

  public let stackSize: Int
  public let used: Int
  public let available: Int
  init() {
    let thread = pthread_self()
    let stackSize = pthread_get_stacksize_np(thread)
    var used: Int = 0
    withUnsafeMutablePointer(to: &used) {
      let stackAddress = Int(bitPattern: pthread_get_stackaddr_np(thread))
      $0.pointee = stackAddress - Int(bitPattern: $0)
    }
    self.stackSize = stackSize
    self.used = used
    self.available = stackSize - used
  }

  public var usedFraction: Double { Double(used) / Double(stackSize) }
  public var availableFraction: Double { 1 - usedFraction }

  public var debugDescription: String {
    return """
      Stack state: \
      using \(used) bytes out of \(stackSize), \
      \(String(format: "%.2f%%", usedFraction * 100)) full.
      """
  }
}

extension Reducer {
  public func debugStack() -> Self {
    var didPrintSize: Bool = false
    return self.combined(with: Reducer<State, Action, Environment> { state, _, _ in
      if !didPrintSize {
        print("\(State.self) size: \(MemoryLayout.size(ofValue: state)) bytes.")
        didPrintSize = true
      }
      StackState.print()
      return .none
    })
  }
}
```

**Reply by tgrapperon** [reply]
@mbrandonw, @stephencelis, if this thing works and is fast enough, maybe we could consider emitting runtime warnings if certain thresholds are reached? Like if we reach 90% of the stack's size for example. If the app crashes because of a stack overflow, there will be at least a trace that could provide some insight and workarounds to the user.

**Reply by tgrapperon** [reply]
I've drafted #1152 following this idea. If someone having this kind of issue could check my branch at `https://github.com/tgrapperon/swift-composable-architecture/tree/runtime-stack-checks`, that would be great! Hopefully, the app should present runtime warnings (the purple ones) before crashing. If you're using SwiftPM to import TCA, you can use:
```swift
.package(url: "https://github.com/tgrapperon/swift-composable-architecture", branch: "runtime-stack-checks")
```
instead to directly check the effect (you will probably need to reset your package cache).

**Reply by tianmingxu-sq** [reply]
>  ```
> public struct StackState: CustomDebugStringConvertible {
>   public static var current: Self { .init() }
>   public static func print() { Swift.print(current) }
> 
>   public let stackSize: Int
>   public let used: Int
>   public let available: Int
>   init() {
>     let thread = pthread_self()
>     let stackSize = pthread_get_stacksize_np(thread)
>     var used: Int = 0
>     withUnsafeMutablePointer(to: &used) {
>       let stackAddress = Int(bitPattern: pthread_get_stackaddr_np(thread))
>       $0.pointee = stackAddress - Int(bitPattern: $0)
>     }
>     self.stackSize = stackSize
>     self.used = used
>     self.available = stackSize - used
>   }
> 
>   public var usedFraction: Double { Double(used) / Double(stackSize) }
>   public var availableFraction: Double { 1 - usedFraction }
> 
>   public var debugDescription: String {
>     return """
>       Stack state: \
>       using \(used) bytes out of \(stackSize), \
>       \(String(format: "%.2f%%", usedFraction * 100)) full.
>       """
>   }
> }
> ```

Hi folks, late to the party but wanna ask a quick question around this tool. We are investigating a stackoverflow issue recently [this issue](https://github.com/pointfreeco/swift-composable-architecture/pull/3452) using this diagnose tool, and I noticed that when printing out the size of the State during state change, even if the state has differed a lot, the state remains the same all the time:
```
Current stack status
- Stack usage: "Stack state: using 145488 bytes out of 8372224, 1.74% full."
- Stack size: "State size: 2016 bytes."
```

Wondering if this is usually the case when there isn't any recursing effects on actions. Thanks for all the help 🙇

**Reply by stephencelis** [reply]
The size in memory of state should remain relatively constant. It's introducing heap allocations like copy-on-write wrapper classes and indirect enums that should bring the size down.

**Comment by bioche** [comment]
Our team is also facing the same issue and I just stumbled upon [this bit of documentation](https://developer.apple.com/library/archive/qa/qa1419/_index.html) that explains how to increase stack size (credit to [this StackOverflow](https://stackoverflow.com/questions/7509729/increase-the-stack-size-of-application-in-xcode-4) (which name is on point in this context ^^))

> Add -Wl,-stack_size,1000000 to the Other Linker Flags field of the Build Styles pane.

It seemed a bit too good to be true but it appears to work according to my first tests using @tgrapperon nice utility above. Stack size apparently went from 1032192 to 16777216 when running on iPhone 11, iOS 15.5. 
 My only fear is now that it would cause app rejection on store. I'll post updates here in any case :)

It feels like a quick patch but it will give us some air until we can solve this in a more futureproof manner ;)

**Reply by DaanyalMeyer** [reply]
> Our team is also facing the same issue and I just stumbled upon [this bit of documentation](https://developer.apple.com/library/archive/qa/qa1419/_index.html) that explains how to increase stack size (credit to [this StackOverflow](https://stackoverflow.com/questions/7509729/increase-the-stack-size-of-application-in-xcode-4) (which name is on point in this context ^^))
> 
> 
> 
> > Add -Wl,-stack_size,1000000 to the Other Linker Flags field of the Build Styles pane.
> 
> 
> 
> It seemed a bit too good to be true but it appears to work according to my first tests using @tgrapperon nice utility above. Stack size apparently went from 1032192 to 16777216 when running on iPhone 11, iOS 15.5. 
> 
>  My only fear is now that it would cause app rejection on store. I'll post updates here in any case :)
> 
> 
> 
> It feels like a quick patch but it will give us some air until we can solve this in a more futureproof manner ;)

We had to solve this issue by breaking the recursive navigation and creating a new store: 

https://github.com/pointfreeco/swift-composable-architecture/issues/620#issuecomment-1207216342

**Reply by bioche** [reply]
Thank you for mentioning that ! Separating stores is a possibility but we still would have to handle the communication between those manually, so I'm not sur it's something we want to explore until we've exhausted all our possibilities :)

**Reply by bioche** [reply]
Unfortunately this trick was indeed too good to be true and seems to only work when deploying directly from xcode as it doesn't seem to survive the archiving phase :/

**Reply by mbrandonw** [reply]
We’ve made a few improvements to the library that should help with stack sizes. We haven’t released them yet, but the changes are on main. Could you point your project to main and see if it helps?

Also the changes in the reducer protocol beta should help a lot, but that will require you to convert most of your reducers to the protocol style to get the most benefit.

**Reply by bioche** [reply]
Thanks for the update @mbrandonw. We are currently still on a Rx based version of the composable architecture but looking forward to switching to the official version soon ;) For now we are managing using @CopyOnWrite modifiers on large substates :)

**Reply by mbrandonw** [reply]
Ohh, sorry I missed that. Hope you're able to work around. 🙂

**Reply by bioche** [reply]
don't worry you didn't miss it I didn't mention it actually ^^ For now it's managable this way even if the app will still crash if user stacks a large number of screens as we have recursive states. Switching the "official" TCA will be the next big task for me :)
